lib/Sema/ConstraintGraph.cpp

```diff
@@ -680,8 +680,34 @@ bool ConstraintGraph::contractEdges() {
         // it's use in the body. If parameter declaration is bound to
         // `inout` it's use has to be bound to `l-value`, which can't
         // happen once equivalence classes of parameter and argument are merged.
-        if (isParamBindingConstraint && tyvar1->getImpl().canBindToInOut())
-          continue;
+        if (isParamBindingConstraint && tyvar1->getImpl().canBindToInOut()) {
+          bool isNotContractable = true;
+          // If we can prove that there are no possible bindings
+          // which result in attempt to bind `inout` type to parameter
+          // type variable, we should go ahead and allow (temporary)
+          // contraction because that greatly helps with performance.
+          if (auto bindings = CS.getPotentialBindings(tyvar1)) {
+            for (auto &binding : bindings.Bindings) {
+              auto type = binding.BindingType;
+              isNotContractable = type.findIf([&](Type nestedType) -> bool {
+                if (auto tv = nestedType->getAs<TypeVariableType>()) {
+                  if (!tv->getImpl().mustBeMaterializable())
```

Comment:
I'm not sure I follow what you're doing here. The bindings we get back from getPotentialBindings() are possibly an incomplete set, so it's not clear to me that we can use them for making a choice like this in general.

Reply:
I think this is valid to do in this specific case in relation to left-hand side of the `bind param` constraint, because it can only take bindings when overload is applied, which means it's either not going to have any and we don't contract, or we have something and we can check if it can possibly be `inout` in the future or not.